pdns/misc.cc

```diff
@@ -510,17 +510,14 @@ string urlEncode(const string &text)
   return ret;
 }
 
-string getHostname()
+std::optional<string> getHostname()
 {
-#ifndef MAXHOSTNAMELEN
-#define MAXHOSTNAMELEN 255
-#endif
-
-  char tmp[MAXHOSTNAMELEN];
-  if(gethostname(tmp, MAXHOSTNAMELEN))
-    return "UNKNOWN";
+  std::array<char, HOST_NAME_MAX> tmp{};
```

Comment:
From http://man.openbsd.org/gethostname:
Note that the corresponding POSIX definition HOST_NAME_MAX in <limits.h> does not
include the terminating NUL character.

So both occurences of `HOST_NAME_MAX` should be `HOST_NAME_MAX + 1`

Reply:
Or rather use getMaxHostNameSize()
See also https://docs.powerdns.com/recursor/security-advisories/powerdns-advisory-2020-03.html

Reply:
I'm not sure `getMaxHostNameSize()` can be used in a const context. It also - I think - isn't adding space for the NUL byte. I changed the code to use `MAXHOSTNAMELEN` and define it to 256 if it is not defined in accordance with the OpenBSD manpage: `Host names are limited to MAXHOSTNAMELEN (from <sys/param.h>) characters, currently 256. This includes the terminating NUL character.`

Reply:
> I'm not sure `getMaxHostNameSize()` can be used in a const context.

It cannot, but you could use a vector instead of an array.

>It also - I think - isn't adding space for the NUL byte.

You need to add 1 indeed.

>I changed the code to use `MAXHOSTNAMELEN` and define it to 256 if it is not defined in accordance with the OpenBSD manpage: `Host names are limited to MAXHOSTNAMELEN (from <sys/param.h>) characters, currently 256. This includes the terminating NUL character.`

The problem with `MAXHOSTNAMELEN` is that it is a non-posix define. So the chances of that name being available are even less than `HOST_NAME_MAX`. I really would use `getMaxHostNameSize()` as it is proven to work on many (even exotic) systems.